### ISSUE TITLE
How to visualize Figure 7?

### ISSUE THREAD

**Issue by Solo4working** [bo-miao/SgMg#6]
Hi, bo-miao.

Thank you for such a great job. 🎉🎉

I am very curious about how the picture you drew in Figure 7 visualizes this heatmap. Could you share the related code? This may be of great help in understanding the working mechanism of sgmg, and even in the field of RVOS. 

Looking forward to your reply~

<img width="516" alt="image" src="https://github.com/bo-miao/SgMg/assets/138687643/2940663e-6e42-4e27-a8d0-464d6eeb4824">

**Comment by bo-miao** [comment]
Hi,

I did not save the original code for this part but rewrote a demo for your reference.

` 

    class InferenceCore:
        def __init__(self, net):
            self.net = net
            self.activation = {}
            self.net.decoder.up_16_8.register_forward_hook(self.get_activation('decoder1'))
            self.net.decoder.up_8_4.register_forward_hook(self.get_activation('decoder2'))
    
        def get_activation(self, name):
            def hook(model, input, output):
                self.activation[name] = output.detach().cpu()
            return hook
    
        def forward(self, inp):
            enc_feat = self.net.forward(inp)
    
            keys = self.activation.keys()
            for key in keys:
                value = self.activation[key]
                if len(value.size()) == 4:
                    value = value[0]
                    c, h, w = value.size()
                    value = torch.sum(value, 0) / c  # * -1 in some networks
                    fig = plt.figure(figsize=(h, w))
                    imgplot = plt.imshow(value)
                    plt.axis("off")
                    plt.savefig('yourpath.jpg', bbox_inches='tight')`

**Comment by Solo4working** [comment]
> Hi,
> 
> I did not save the original code for this part but rewrote a demo for your reference.
> 
> `
> 
> ```
> class InferenceCore:
>     def __init__(self, net):
>         self.net = net
>         self.activation = {}
>         self.net.decoder.up_16_8.register_forward_hook(self.get_activation('decoder1'))
>         self.net.decoder.up_8_4.register_forward_hook(self.get_activation('decoder2'))
> 
>     def get_activation(self, name):
>         def hook(model, input, output):
>             self.activation[name] = output.detach().cpu()
>         return hook
> 
>     def forward(self, inp):
>         enc_feat = self.net.forward(inp)
> 
>         keys = self.activation.keys()
>         for key in keys:
>             value = self.activation[key]
>             if len(value.size()) == 4:
>                 value = value[0]
>                 c, h, w = value.size()
>                 value = torch.sum(value, 0) / c  # * -1 in some networks
>                 fig = plt.figure(figsize=(h, w))
>                 imgplot = plt.imshow(value)
>                 plt.axis("off")
>                 plt.savefig('yourpath.jpg', bbox_inches='tight')`
> ```

Thanks for your quick reply, I will try the code you shared with me to better understand how SgMg works.
